R/dt_spanners.R

```diff
@@ -137,106 +113,80 @@ dt_spanners_print_matrix <- function(
-
-    if (length(vars_spanner_i_flt) < 1) {
-      # If none of the columns under the proposed spanner are
-      # set as visible columns, then we need to mark the spanner
-      # entry for deletion from `spanners_tbl`
-      remove_entry[i] <- TRUE
-    } else {
-      spanners_tbl[i, ][["vars"]] <- list(vars_spanner_i_flt)
-    }
-  }
-
-  # Remove entries from `spanners_tbl` as needed
-  spanners_tbl <- spanners_tbl[!remove_entry, ]
+  spanners_tbl <-
+    spanners_tbl %>%
+    dplyr::mutate(vars = lapply(.data$vars, base::intersect, .env$vars)) %>%
+    dplyr::filter(vapply(vars, length, integer(1)) > 0) %>%
+    dplyr::mutate(
+      spanner_level = match(spanner_level, sort(unique(spanner_level)))
+    )
 
-  # Get the maximum height of the spanner matrix; this won't
-  # necessarily be the finalized height since we will collapse
-  # empty rows
+  # If `spanners_tbl` is immediately empty then return a single-row
+  # matrix of column vars/IDs (or not, if `omit_columns_row = TRUE`)
   if (nrow(spanners_tbl) < 1) {
-    spanner_height <- 0
-  } else {
-    spanner_height <- max(spanners_tbl[["spanner_level"]])
+    return(
+      empty_spanner_matrix(vars = vars, omit_columns_row = omit_columns_row)
+    )
   }
 
+  # Get the height of the spanner matrix
+  spanner_height <- max(spanners_tbl[["spanner_level"]])
+
   vars_vec <- rep(NA_character_, length(vars))
   names(vars_vec) <- vars
 
-  # Initialize matrix to serve as boxhead representation
-  columns_mat <- matrix(vars, nrow = 1, ncol = length(vars))
+  # Initialize matrix row that represents the column vars
+  columns_mat_cols_row <- matrix(vars, nrow = 1, ncol = length(vars))
 
+  # Initialize matrix to serve as boxhead representation
   columns_mat <-
-    rbind(
-      columns_mat,
-      if (spanner_height > 0) {
-      matrix(vars_vec, nrow = spanner_height, ncol = length(vars_vec))
-      }
-    )
+    matrix(NA_character_, nrow = spanner_height, ncol = length(vars_vec))
   colnames(columns_mat) <- vars
 
-  spanners_tbl <- spanners_tbl %>% dplyr::filter(spanner_level >= 1)
-
-  columns_mat_id <- columns_mat_labels <- columns_mat
-
-  if (spanner_height > 0) {
-
-    for (i in seq_len(nrow(spanners_tbl))) {
+  # columns_mat_id <- columns_mat_labels <- columns_mat
 
-      columns_mat_id[spanners_tbl$spanner_level[[i]] + 1, spanners_tbl$vars[[i]]] <-
-        spanners_tbl$spanner_id[i]
-
-      columns_mat_labels[spanners_tbl$spanner_level[[i]] + 1, spanners_tbl$vars[[i]]] <-
-        spanners_tbl$built[i]
-    }
-
-    # Flip matrix in y direction to put boxhead levels in display order
-    columns_mat_id <- columns_mat_id[rev(seq_len(nrow(columns_mat_id))), ]
-    columns_mat_labels <- columns_mat_labels[rev(seq_len(nrow(columns_mat_labels))), ]
+  for (i in seq_len(nrow(spanners_tbl))) {
+    columns_mat[spanners_tbl$spanner_level[[i]], spanners_tbl$vars[[i]]] <-
+      if (ids) spanners_tbl$spanner_id[i] else spanners_tbl$built[i]
```

Comment:
For consistency with previous line
```suggestion
      if (ids) spanners_tbl$spanner_id[[i]] else spanners_tbl$built[[i]]
```

Reply:
Now changed.